test/parallel_api/algorithm/alg.modifying.operations/shift_left_right.pass.cpp

```diff
@@ -0,0 +1,209 @@
+            res_idx = het_res - ptr.get();
+
+            //3.2 check result
+            algo.check(ptr.get() + res_idx, ptr.get(), m, first_exp, n);
+        }
+#endif
+    }
+#endif
+};
+
+template <typename T, typename Size, typename Algo>
+void
+test_shift_by_type(Size m, Size n, Algo algo)
+{
+    TestUtils::Sequence<T> orig(m, [](::std::size_t v) -> T { return T(v); }); //fill data
+    TestUtils::Sequence<T> in(m, [](::std::size_t v) -> T { return T(v); }); //fill data
+
+    TestUtils::invoke_on_all_policies<>()(test_shift(), in.begin(), m, orig.begin(), n, algo);
+}
+
+struct shift_left_algo
+{
+    template <typename Policy, typename It>
+    It operator()(Policy&& exec, It first, It last, typename ::std::iterator_traits<It>::difference_type n)
+    {
+        return oneapi::dpl::shift_left(::std::forward<Policy>(exec), first, last, n);
+    }
+
+    template <typename It, typename ItExp>
+    void
+    check(It res, It first, typename ::std::iterator_traits<It>::difference_type m, ItExp first_exp,
+        typename ::std::iterator_traits<It>::difference_type n)
+    {
+        //if (n > 0 && n < m), returns first + (m - n). Otherwise, if n  > 0, returns first.
+        //Otherwise, returns last.
+        It __last = ::std::next(first, m);
+        auto res_exp = (n > 0 && n < m ? ::std::next(first, m - n) : (n > 0 ? first : __last));
+
+        EXPECT_TRUE(res_exp == res, "wrong return value of shift_left");
+
+        if(res != first && res != __last)
+        {
+            EXPECT_EQ_N(first, ::std::next(first_exp, + n), m - n, "wrong effect of shift_left");
+            //restore unput data
+            std::copy_n(first_exp, m, first);
+        }
+    }
+};
+
+struct shift_right_algo
+{
+    template <typename Policy, typename It>
+    typename ::std::enable_if<TestUtils::is_same_iterator_category<It, ::std::bidirectional_iterator_tag>::value
+                            || TestUtils::is_same_iterator_category<It, ::std::random_access_iterator_tag>::value,
+                            It>::type
+    operator()(Policy&& exec, It first, It last, typename ::std::iterator_traits<It>::difference_type n)
+    {
+        return oneapi::dpl::shift_right(::std::forward<Policy>(exec), first, last, n);
+    }
+    //skip the test for non-bidirectional iterator (forward iterator, etc)
+    template <typename Policy, typename It>
+    typename ::std::enable_if<!TestUtils::is_same_iterator_category<It, ::std::bidirectional_iterator_tag>::value
+                            && !TestUtils::is_same_iterator_category<It, ::std::random_access_iterator_tag>::value,
+                            It>::type
+    operator()(Policy&& exec, It first, It last, typename ::std::iterator_traits<It>::difference_type n)
+    {
+        return first;
+    }
+
+    template <typename It, typename ItExp>
+    typename ::std::enable_if<TestUtils::is_same_iterator_category<It, ::std::bidirectional_iterator_tag>::value
```

Comment:
Please investigate the possibility to change this check to `is_base_of` semantics (after merging this review)